### PR TITLE
P2443R1 views::chunk_by

### LINKED ISSUE #5272: [2022-02 LWG Motion 13] P2443R1 views::chunk_by
[P2443R1](https://wg21.link/p2443r1)

### PR DESCRIPTION
**Pull request by jensmaurer**
Fixes #5272
Fixes cplusplus/papers#1109

### CODE REVIEW

File: source/ranges.tex [hunk 1]
```diff
@@ -10040,3 +10048,330 @@
...
 Equivalent to \tcode{return x.\exposid{inner_} - y.\exposid{inner_};}
 \end{itemdescr}
+
+\rSec2[range.chunk.by]{Chunk by view}
+
+\rSec3[range.chunk.by.overview]{Overview}
+
+\pnum
+\tcode{chunk_by_view} takes a \libconcept{view} and a predicate, and
+splits the \libconcept{view} into subranges
```
Comment (JohelEGP) [inline source/ranges.tex:10058]: This is what the paper suggests, and is used elsewhere:
> splits the \libconcept{view} into \tcode{subrange}s on the delimiter.
```suggestion
splits the \libconcept{view} into \tcode{subrange}s
```

File: source/ranges.tex [hunk 2]
```diff
@@ -10040,3 +10048,330 @@
...
+  cout << '[';
+  auto sep = "";
+  for(auto i : r) {
+    cout << sep << i;
+    sep = ", ";
+  }
+  cout << "] ";
+}
+// The above prints: [1, 2, 2, 3] [0, 4, 5] [2]
+\end{codeblock}
```
Comment (JohelEGP) [inline source/ranges.tex:10083]:
```suggestion
\end{codeblock}
The above prints: \tcode{[1, 2, 2, 3] [0, 4, 5] [2]}
```

File: source/ranges.tex [hunk 3]
```diff
@@ -10040,3 +10048,330 @@
...
+\end{example}
+
+\rSec3[range.chunk.by.view]{Class template \tcode{chunk_by_view}}
+
+\begin{codeblock}
+namespace std::ranges {
+  template<@\libconcept{forward_range}@ V, @\libconcept{indirect_binary_predicate}@<iterator_t<V>, iterator_t<V>> Pred>
+    requires @\libconcept{view}@<V> && is_object_v<Pred>
+  class chunk_by_view : public view_interface<chunk_by_view<V, Pred>>{
```
Comment (JohelEGP) [inline source/ranges.tex:10092]:
```suggestion
  class chunk_by_view : public view_interface<chunk_by_view<V, Pred>> {
```

File: source/ranges.tex [hunk 4]
```diff
@@ -10040,3 +10048,330 @@
...
+
+\begin{itemdecl}
+constexpr explicit chunk_by_view(V base, Pred pred);
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Initializes \exposid{base_} with \tcode{std::move(base)} and
+@\exposid{pred_}@ with \tcode{std::move(pred)}.
```
Comment (JohelEGP) [inline source/ranges.tex:10129]:
```suggestion
\exposid{pred_} with \tcode{std::move(pred)}.
```

File: source/ranges.tex [hunk 5]
```diff
@@ -10040,3 +10048,330 @@
...
+\end{itemdescr}
+
+\begin{itemdecl}
```
Comment (JohelEGP) [inline source/ranges.tex:10132]:
```suggestion
\indexlibrarymember{pred}{chunk_by_view}%
\begin{itemdecl}
```

File: source/ranges.tex [hunk 6]
```diff
@@ -10040,3 +10048,330 @@
...
+\end{itemdecl}
+
+\begin{itemdescr}
+\pnum
+\effects
+Equivalent to: \tcode{return *\exposid{pred_};}
+\end{itemdescr}
+
+\begin{itemdecl}
+constexpr iterator begin();
```
Comment (JohelEGP) [inline source/ranges.tex:10143]:
```suggestion
constexpr @\exposid{iterator}@ begin();
```

File: source/ranges.tex [hunk 7]
```diff
@@ -10040,3 +10048,330 @@
...
+\tcode{\exposid{pred_}.has_value()} is \tcode{true}.
+
+\pnum
+\returns
+\tcode{\exposid{iterator}(*this, ranges::begin(\exposid{base_}), \exposid{find-next}(ranges::begin(\exposid{base_})))}.
+
+\pnum
+\remarks
+In order to provide
+the amortized constant-time complexity required by the range concept,
```
Comment (JohelEGP) [inline source/ranges.tex:10158]:
```suggestion
the amortized constant-time complexity required by the \libconcept{range} concept,
```

File: source/ranges.tex [hunk 8]
```diff
@@ -10040,3 +10048,330 @@
...
+\begin{itemize}
+\item
+\tcode{current} is not equal to \tcode{ranges::begin(\exposid{base_})}.
+\item
+\tcode{\exposid{pred_}.has_value()} is \tcode{true}.
+\end{itemize}
+
+\pnum
+\returns
+An \exposid{iterator} \tcode{i}
```
Comment (JohelEGP) [inline source/ranges.tex:10212]:
```suggestion
An iterator \tcode{i}
```

File: source/ranges.tex [hunk 9]
```diff
@@ -10040,3 +10048,330 @@
...
+in the range \range{ranges::begin(\exposid{base_})}{current} such that:
+\begin{itemize}
+\item
+\tcode{ranges::adjacent_find(i, current, not_fn(ref(*\exposid{pred_})))} is equal to \exposid{current}; and
```
Comment (JohelEGP) [inline source/ranges.tex:10216]:
```suggestion
\tcode{ranges::adjacent_find(i, current, not_fn(ref(*\exposid{pred_})))} is equal to \tcode{current}; and
```

**Review by JohelEGP — CHANGES_REQUESTED** [review]

File: source/ranges.tex (hunk 5 above)
Comment (JohelEGP) [reply source/ranges.tex:10133]: I just happened to notice that all other `pred` were indexed like this.

File: source/ranges.tex [hunk 10]
```diff
@@ -9523,7 +9530,8 @@
 \begin{itemdescr}
 \pnum
 \effects
-Equivalent to: \tcode{return y.\exposid{end_} - x.\exposid{current_}.back();}
+Equivalent to:
```
Comment (tkoeppe) [inline source/ranges.tex:9533]: @jensmaurer: What is the purpose of this edit?

**Pull request merged**